carbon-design-system/carbon
# fix(typescript): rename index.js to index.ts to expose some components

Closes #14165 

#### Changelog

**Changed**

- Rename several `index.js` files to `index.ts` to make sure export statements in `node_modules/@carbon/react/lib/index.d.ts` of a TS React project can be resolved by the TypeScript compiler. (Reference: [How TypeScript resolves modules](https://www.typescriptlang.org/docs/handbook/module-resolution.html#how-typescript-resolves-modules))

(Note: This PR renames `index.js` files that only reference TS files. There are still many components that cannot be imported in a TS project. Fixing the problem for all components in a single PR may lead to a huge PR which would be hard to test and review, so I only fix the easiest ones in this PR.)

#### Testing / Reviewing

- CI should pass.
- Existing stories should work the same as before.
- Manually verify the library built on the new code exposes components like `DatePickerInput` by following the steps below.
  - Build the project: `yarn build`
  - Create a new TypeScript React project: `npx create-react-app test-carbon-import-ts --template typescript`
  - In the newly created project, install `@carbon/react`: `npm i @carbon/react`
  - Copy the built lib to the new project (replace the path to the path on your machine): `cp -r /root/my-fork/carbon/packages/react/lib /root/tmp/test-carbon-import-ts/node_modules/@carbon/react/`
  - Add the following code snippet to `src/App.tsx` in the new project and verify those components are imported correctly.
```ts
import {
  ComboBox,
  DatePickerInput,
  ExpandableSearch,
  FileUploader,
  Filename,
  FileUploaderSkeleton,
  FileUploaderButton,
  FileUploaderDropContainer,
  FileUploaderItem,
  FormGroup,
  Loading,
  NumberInputSkeleton,
  NumberInput,
  OverflowMenuItem,
  RadioButton,
  RadioButtonGroup,
  SelectSkeleton,
  Select,
  SelectItem,
  SelectItemGroup,
  TimePicker,
  TimePickerSelect,
} from "@carbon/react";
```
